tests/Regressions.hs

```diff
@@ -71,6 +76,57 @@ propEqAfterDelete (Keys keys) =
 mapFromKeys :: [Int] -> HM.HashMap Int ()
 mapFromKeys keys = HM.fromList (zip keys (repeat ()))
 
+------------------------------------------------------------------------
+-- Issue #254
+
+-- Key type that always collides.
+newtype KC = KC Int
+  deriving (Eq, Ord, Show)
+instance Hashable KC where
+  hashWithSalt salt _ = salt
+
+issue254Lazy :: Assertion
+issue254Lazy = issue254LazyLambda 2
+
+-- We want to make sure that old values in the HashMap are evicted when new values are inserted,
+-- even if they aren't evaluated. To do that, we use the WeakPtr trick described at
+-- http://simonmar.github.io/posts/2018-06-20-Finding-fixing-space-leaks.html.
+-- We insert a value named oldV into the HashMap, then insert over it, checking oldV is no longer reachable.
+--
+-- To make the test robust, it's important that oldV isn't hoisted up to the top or shared. To do that,
+-- we use NOINLINE, make oldV dependent on an unseen argument, and insert _ <- return () to ensure oldV
+-- is under a lambda.
```

Comment:
Yuck yuck yuck. I don't like this approach to avoiding floating. It seems much too fragile. Fortunately, I think we can do something both simpler and more robust, taking advantage of the surprising fact that `type Assertion = IO ()`.

```haskell
issue254Lazy :: Assertion
issue254Lazy = do
  i :: Int <- randomIO
  let oldV = error $ "...." ++ show i
  ....
```

Now GHC has no way to float out the old value because it's genuinely unknown until the assertion is run.

Reply:
That's way better. I've done that.